### DISCUSSION TITLE
Getting/Setting consumer state

### DISCUSSION THREAD

**Discussion by matan129** [confluentinc/parallel-consumer#815] [Q&A]
Hi,

One of my apps uses Parallel Consumer 0.5.2.8.
It consumes messages from Kafka with consumer group A, and maintains some data structure on-disk.

This data structure (well - a collection of files) is periodically backed up to remote storage. 

I have a use case in which a different service (with consumer group B), needs to download the backup and start ingesting from the point that service A left it.

**Q**: Is there a way to get (and set?) the current position & state of the consumer?

Currently, I get the committed offsets per partition and use it as an approximation, but it's not granular enough.

Thanks!

**Comment by astubbs** [comment]
@matan129 - there is no live getter or setter for the consumer's position, but the state you want is already in Kafka, and it is more granular than the committed offset alone.

PC's durable state per partition is exactly the committed `OffsetAndMetadata`: the offset is the lowest *incomplete* record, and the metadata string is an encoded map of every completed offset above it (README, "Offset Map" and "Storage Notes"). That pair is the whole of what PC restores on assignment, so it is the granular position you are after, and `AdminClient.listConsumerGroupOffsets` on group A already returns both halves. Seeding group B with the same pair is possible in principle, since PC decodes whatever metadata it finds on assignment, but there is no supported API for it and I have not run that route, so I will not describe it as one.

Where the fork stands: the state itself has not changed on [v0.6.0.0](https://github.com/astubbs/parallel-consumer/releases/tag/v0.6.0.0) beyond hardening the decoder against corrupt metadata. A readable, settable position is the open issue astubbs/parallel-consumer#157; if this backup-and-resume case is what you need, describe it there and it can be made an API rather than a recipe. Two open pull requests, not merged, sit beside it: a health-check surface that reports run state and failure cause, not positions (astubbs/parallel-consumer#226), and an embedded dashboard showing the in-flight state (astubbs/parallel-consumer#268).

The coordinates changed with the fork - on Maven Central it is now:

```xml
<dependency>
    <groupId>bz.stub.parallelconsumer</groupId>
    <artifactId>parallel-consumer-core</artifactId>
    <version>0.6.0.0</version>
</dependency>
```

(`parallel-consumer-vertx`, `parallel-consumer-reactor` and `parallel-consumer-mutiny` likewise.) The Java package is `bz.stub.parallelconsumer` too, so imports change with it - one command rewrites them (GNU `sed`; on macOS use `sed -i ''`):

```bash
find . -name '*.java' -exec sed -i 's/io\.confluent\.parallelconsumer/bz.stub.parallelconsumer/g' {} +
```
